### ISSUE TITLE
Dev: Implement Guest Workflow Tasks Endpoints

### ISSUE THREAD

**Issue by ju1es** [hackforla/HomeUniteUs#572]
Note: added draft label until dev team is ready to identify next steps for this issue. 9/12/24 Ariel Lasry

### Overview
The API shall manage the HUU housing Workflow. API clients (e.g. the front-end application) should be able to view Tasks (name, description, status) assigned to Guests by the Workflow. The API clients should also allow Guests to navigate to Tasks to view and work on.

Tasks are an abstract concept and their concrete implementations represent such things as Forms, Training/Lessons, Scheduling, Matching, etc. The Tasks in the Workflow are logically grouped together into Task Groups. The Task Groups represent major sub-processes of the Workflow such as "Application & Onboarding process", "Host Matching process", "Match Finalization process", etc. 

For this issue, all Task Groups and Tasks up until the matching process should be returned by the API.

Task Groups and Tasks go through phases represented by the following statuses:
1. Locked
2. Start
3. In progress
4. More information is needed
5. Complete

The API shall maintain the following invariants imposed by the Workflow:

1. The order of the Task Groups represents the order in which the Guest must complete the Task Groups.
2. The order of the Tasks represents the order in which the Guest must complete the Tasks within a Task Group.
3. The Tasks in a Task Group can not be unlocked until the Task Group is unlocked and the Task before it has been completed. If it's the first Task in the Task Group, then it is automatically unlocked only if the Task Group is unlocked.
6. A Task Group is completed only when all of its Tasks are completed. The next Task Group can then be unlocked.

### Action Items
Domain concerns:
- [ ] Implement object model for Guests, Task Groups, Tasks, Workflow and their associations
- [ ] Implement state machine representing the (Guest) Workflow

The following data must be available in the endpoint(s):
- [ ] Guest Task Groups and their Tasks up until the matching process.
- [ ] Task Group ID, Title, Status, and Tasks.
- [ ] Task ID, Title, Description, Status, Link Text, Link URL Path

Database concerns:
- [ ] Implement a persistence layer retrieve, store, and update Tasks
- [ ] Remove unused tables
- [ ] Implement relational model for Task Groups, Tasks, Workflow and associations with Guests
- [ ] Implement migration script(s)

### Resources/Instructions
- [Section 2 Epic](https://github.com/hackforla/HomeUniteUs/issues/500)
- [Section 2 Guest Epic Google Doc](https://docs.google.com/document/d/141qIw7FDyk3BoVAfc7iBsdhMqtLKp_y5MA082ouc2Fk/edit)
- [Frontend Work](https://github.com/hackforla/HomeUniteUs/issues/566)
- [Frontend Designs](https://www.figma.com/file/BNWqZk8SHKbtN1nw8BB7VM/HUU-Everything-Figma-Nov-2022?type=design&node-id=8700%3A7901&mode=design&t=RYccU4PYyIHb5Qe7-1)
- [Workflow on Figma](https://www.figma.com/file/BNWqZk8SHKbtN1nw8BB7VM/HUU-Everything-Figma-Nov-2022?type=design&node-id=2353-17312&mode=design&t=EcEz499oYWUpyIyw-0)
- [State machine issue](https://github.com/hackforla/HomeUniteUs/issues/391)

**Comment by Joshua-Douglas** [comment]
# Overview

Create the endpoints needed by the frontend application to create and save the guest tasks. The guest onboarding process requires the guest to complete multiple tasks. Each task will have some associated data that needs to be stored. The onboarding is considered complete once all tasks are complete.

The backend should store the overall progress of the guest application, along with the detailed information associated with each task.

## Requirements

In this initial PR, we will only store completed tasks. We can introduce draft task persistance in the future, this PR will be complex enough without this additional feature.

* Store the guest's main task progress in the HUU database
* Store the information associated with each guest task
* Add functionality needed to delete the guest tasks

## Research Questions

### List the tasks that we need to store in the database

> I count three.
> https://github.com/hackforla/HomeUniteUs/pull/596

### List the data we need to store for each task

> I don't think we know all of this information yet. I suggest we start with copying all relevant questions from this application <https://docs.google.com/forms/d/e/1FAIpQLSc2Zm709r_7avFQYcaL9pZRwAnUknCZengn8rXP6jxx3sm9vQ/viewform?gxids=7757>.

#### Data for task 1

> List the data needed for the task, along with the data types. This will form the JSON schema that is used to send/receive data from the backend

#### Data for task 2
#### etc.

### Database

#### Do user accounts have roles associated with them?

> Hint: No (See the User class). Reading the data models is currently more challenging then it needs to be because we have a lot of unused models. You need to refer to the endpoints to see which are being used.
> Explain why this is a problem

#### How do you define a new data model?

> Show me an example model that has a bool, str, primary key, and date. SQLAlchemy data models can get quite advanced so let's start simple.

#### What are data models used for?

> SQLAlchemy docs will be useful here. How does the database use the data model?

#### How do you use the new data model to actually add an entry to the database?

> Show me how to add the datamodel you defined up above

#### How do you retrieve data models from the database

> Show me how to get back the data model you added in previous step

#### How do you define a relationship between two tables in a database using SQLAlchemy

> This is an advanced question! But we will need to associate guests with their tasks

#### How do you convert a data model into JSON?

> Hint: We use a library to do this. SQLAlchemy does not provide this functionality.

### Endpoints

#### Where are endpoints located within the API code?

> Show me one of our endpoints here. Tell me what information is entering the endpoint and what information is leaving, in detail.

#### Describe the process for adding a new endpoint

> Start simple. Describe how to add an endpoint that returns a Guest using their email. This document and the associated repo should give you a head start: https://github.com/Joshua-Douglas/PythonRestAPIExamples/blob/main/OpenAPI_API_Tools.md

### Testing

#### How do you run the backend test cases?

> Be brief. Just want to make sure you know how to setup and run them on your machine

#### What is a pytest fixture?

> Show me how to use one. Write a super simple trivial test case using a fixture.

#### Show a test case that adds a value to the database and checks it

> Hint: We already have some test cases that do this. Just show one of these and explain it.

### Show a test case that tests an endpoint

> We already have this as well. Just show a simple one and explain it.

### General Solution Plan

> We will create this once research is complete.

**Comment by Joshua-Douglas** [comment]
Hey @agosmou,

An initial project plan is ready for this issue. Please submit your answers and reassign me once it is ready for review. 

My comments are included as 

> quoted text

Please don't remove those section. I'll remove them as I review your responses.

**Comment by agosmou** [comment]
Working through this on markdown doc in VSCode. Ill paste it in here when I finish over the next day or so.

**Comment by erikguntner** [comment]
@agosmou @Joshua-Douglas I simplified the dashboard data a bit and wanted to leave an example here for reference since I'll be closing out issue 596 soon. 
Sample response for tasks and sub-tasks:

**Comment by agosmou** [comment]
@erikguntner 

Awesome info. Thanks Eric! I'm wrapping up the design doc. I'll post this afternoon.

**Comment by erikguntner** [comment]
@agosmou this looks awesome! Just FYI Cognito tracks whether the user is confirmed in case that changes whether you want to keep that info in the database as well or not

**Comment by Joshua-Douglas** [comment]
Hey @agosmou, Thanks for the designs! I'll make sure to review them by Wednesday night!

**Comment by agosmou** [comment]
Figma Designs for backend-design consideration

@Joshua-Douglas - I can review these tomorrow night to see if there are any effects on the above

https://www.figma.com/file/BNWqZk8SHKbtN1nw8BB7VM/HUU-Everything-Figma-Nov-2022?type=design&node-id=9669-3122&mode=design&t=oPs2UK9Ee4GFVwK4-0

**Comment by Joshua-Douglas** [comment]
Hey @erikguntner, 

Thanks a lot for the Task/SubTask idea. After reading through @agosmou's research I think a generic approach like this is going to be easier to implement than the more 'naive' approach of defining each application as a separate model. 

It would be great if you could look through the proposed data model & application flowchart I posted above. If the frontend is already relying on the backend to query the dashboard and applications then that gives us the flexibility to define provider-specific and role-specific dashboards. This would mean that each individual provider could define a custom Guest and Host dashboards, complete with custom applications.

**Comment by Joshua-Douglas** [comment]
Hey @agosmou,

The design document is ready for review! Your research looked great, and it inspired me to think up a generic approach that could accommodate the Task/Subtask approach outlined by @erikguntner. 

I left 
> comments 

on some of your responses, and added several new sections (everything after the `Data Model Questions` section). I left one question un-answered, so I could really use your help getting to the bottom of it. 

Can you answer that question, and review the design? I'd like to get your feedback. I'm sure something is missing or could be improved. If you agree with the general approach then I'll create a set of implementation questions that we can use to outline the trickiest parts of the implementation before opening a branch.

**Comment by paulespinosa** [comment]
@agosmou The Guest Dashboard story #500 uses Dashboard -> Steps -> Tasks language to describe the Guest Dashboard (see the "Key Decisions" section). In the diagram provided by @erikguntner's, Task should be "Step" and Sub_Task should be "Task" to align with the story's description of the Guest Dashboard feature. If anyone has found it makes more sense to name these entities differently, loop @sanya301 in to iterate on the language we'll use to describe and implement the feature.

The Dashboard Application Model diagram provided by @Joshua-Douglas is a little bit out of scope for this feature but still eventually necessary. It would need to be modified to associate the Application model with a Task: A Dashboard contains Steps, Steps are a group of Tasks, Task has-a/is-a: Application has-a/is-a Form, Training, Interview, or Match, etc.

Is the above a correct representation of the Guest Dashboard model?

**Comment by Joshua-Douglas** [comment]
Hey @paulespinosa, How is the guest dashboard data model diagram out of scope for the guest dashboard endpoint issue? Do you think we should split the two issues and backlog this one, or do you think that changes to the data model are not strictly necessary?

**Comment by paulespinosa** [comment]
@Joshua-Douglas specifically, the "HUU Application Dashboard Data Model" diagram. It contains models (`Application`, `Question`, `ResponseType`, `Response`, `App_Question_Junction`), that are circled in the image below, that are more representative of an application form and are great candidate models for the issues: [Guest Application Form #533](533) and [Guest application implementation #610](610) and beyond.

![pic-selected-231006-2156-34](https://github.com/hackforla/HomeUniteUs/assets/4141567/402e551b-05f7-4769-89b4-b87f49f0d98a)

_Caveat: I've been using the terminology "Steps" and "Tasks" as defined in the main issue [Guest Dashboard #500](500). However, it appears terminology has evolved from there, so I've pinged Sonya https://github.com/hackforla/HomeUniteUs/issues/500#issuecomment-1751054169 to motivate alignment on how the team describes the Dashboard features._

The [Guest Dashboard #500](500) issue, the parent issue for this issue, Implement Guest Dashboard Endpoints #572, specifies displaying "high-level" information such as status, description, and progress about Steps. It's not asking to display the details of an application (form). When #500 talks about being able to "click into any additional page(s)", it means as a generic action where the details of those landing pages are unspecified and left for implementation in separate issues.

So the question becomes: how do you get the status and progress of a Step? Perhaps, by asking each of its Tasks for their status and progress. How do you get the status and progress of a Task representing, say, an Application? Consider, as an implementation option, defining Task as an interface or abstract base class. Have concrete _Dashboard context representations_ of Tasks, such as an Application, contain the logic to calculate or get the status and progress. For this issue, the concrete implementations can return canned responses until specific issues are defined to implement their details (e.g. regarding Application, it can be an issue on its own or an action item for implementing the backend for [Guest Application Form #533](533)).

Unfortunately, the Action Items in this issue appears to be misaligned with its parent.
![pic-selected-231007-0015-32](https://github.com/hackforla/HomeUniteUs/assets/4141567/c2695de8-b489-42e3-9da7-244f17e9e1cb)
Based on the context of #500, the action items circled above should probably read:
* Steps associated with a Dashboard
* Tasks associated with each Step
* Description of each Step (or Task?) - to be clarified  https://github.com/hackforla/HomeUniteUs/issues/500#issuecomment-1751054169

Regarding #500's Consideration: "In the future, the plan is for the guest onboarding experience for multiple organizations (SPY and beyond), where different organizations might want to modify the steps needed for a guest.". This word "modify" is under-specified and needs to be discussed with Product Management https://github.com/hackforla/HomeUniteUs/issues/500#issuecomment-1751054169.

**Comment by agosmou** [comment]
According to [this comment](https://github.com/hackforla/HomeUniteUs/issues/500#issuecomment-1760713307), we are doing the following naming convention:

"Task Groups has Tasks"

**Comment by agosmou** [comment]
@Joshua-Douglas Take a look at the edits above on the design doc. Let me know if you want to meet to discuss our models further - the ERDiagram you posted puts into perspective the reorganizing that has to be done

**Comment by agosmou** [comment]
Hi @paulespinosa - Are you able to look over my progress setting up models? I want to make sure Im understanding this and going in the right direction.

I tried my hand at what the models would look like so I could then work on the tests, domain objects, and domain logic.

As far as the statemachine (state chart pattern), we will have `models.py` and `statechart.py` that has the logic for this. At this time, I am planning on handrolling unless we agree to using a library.

- Option 1: [python-statemachine](https://github.com/fgmacedo/python-statemachine/blob/develop/pyproject.toml#L27) with support for python 3.10
- Option 2: [transitions](https://github.com/pytransitions/transitions/blob/master/setup.py#L57) with support for up to python 3.8 so Im not sure this can play nice with [our api](https://github.com/hackforla/HomeUniteUs/blob/main/api/pyproject.toml#L17) on python 3.10

**models to be reviewed**
```python
##########################
# api/models/database.py #
##########################
import enum # adds python enum import
from sqlalchemy import create_engine
from sqlalchemy.engine import Engine
from sqlalchemy.orm import Session, declarative_base
from sqlalchemy import Column, Integer, String, ForeignKey, DateTime, LargeBinary, Enum # adds enum
from sqlalchemy.orm import relationship # adds imports

Base = declarative_base()

# proposed models...

class TaskStatus(enum.Enum):
    LOCKED = 'Locked'
    START = 'Start'
    IN_PROGRESS = 'In Progress'
    MORE_INFORMATION_NEEDED = 'More Information Needed'
    COMPLETE = 'Complete'

class TaskGroupStatus(enum.Enum):
    LOCKED = 'Locked'
    START = 'Start'
    IN_PROGRESS = 'In Progress'
    MORE_INFORMATION_NEEDED = 'More Information Needed'
    COMPLETE = 'Complete'

class TaskGroupSubProcess(enum.Enum):
    LOCKED = 'Locked'
    APPLICATION_AND_ONBOARDING_PROCESS = 'Application & Onboarding Process'
    HOST_MATCHING_PROCESS = 'Host Matching Process'
    MATCH_FINALIZATION_PROCESS = 'Match Finalization Process'

class Guest(Base):
    __tablename__ = 'guest'
    id = Column(Integer, primary_key=True, index=True)
    email = Column(String, ForeignKey('user.email'), nullable=False)
    tasks = relationship("Task", back_populates="guest", cascade="all, delete")
    task_groups = relationship("TaskGroup", back_populates="guest", cascade="all, delete")
    guest_workflows = relationship("GuestWorkflow", back_populates="guest", cascade="all, delete")

class Task(Base):
    __tablename__ = 'task'
    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=False)
    description = Column(String, nullable=False)
    task_group_id = Column(Integer, ForeignKey('task_group.id'))
    status = Column(Enum(TaskStatus), default=TaskStatus.LOCKED, nullable=False) # from finite state machine (statechart pattern)
    guest_id = Column(Integer, ForeignKey('guest.id', ondelete='CASCADE'))

    # methods for state machine (statechart pattern)


class TaskGroup(Base):
    __tablename__ = 'task_group'
    id = Column(Integer, primary_key=True, index=True)
    name = Column(Enum(TaskGroupSubProcess), default=TaskGroupSubProcess.LOCKED, nullable=False) # from finite state machine (statechart pattern)
    guest_workflow_id = Column(Integer, ForeignKey('guest_workflow.id'))
    status = Column(Enum(TaskGroupStatus), default=TaskGroupStatus.LOCKED,nullable=False) # from finite state machine (statechart pattern)
    guest_id = Column(Integer, ForeignKey('guest.id', ondelete='CASCADE'))
    guest = relationship("Guest", back_populates="task_groups")
    tasks = relationship("Task", back_populates="task_groups")

    # methods for state machine (statechart pattern)

class GuestWorkflow(Base):
    __tablename__ = 'guest_workflow'
    id = Column(Integer, primary_key=True, index=True)
    guest_id = Column(Integer, ForeignKey('guest.id', ondelete='CASCADE'))
    task_groups = relationship('TaskGroup', back_populates='guest_workflow')
    guest = relationship("Guest", back_populates="guest_workflows")

# existing code..

class User(Base):
    __tablename__ = "user"
    id = Column(Integer, primary_key=True, index=True)
    email = Column(String, nullable=False, unique=True)

class ApplicantType(Base):
    __tablename__ = "applicant_type"

    id = Column(Integer, primary_key=True, index=True)
    applicant_type_description = Column(String, nullable=False)

class Applicant(Base): 
    __tablename__ = "applicant"

    id = Column(Integer, primary_key=True, index=True)
    applicant_type = Column(Integer, ForeignKey('applicant_type.id'), nullable=False)
    applicant_status = Column(Integer, ForeignKey('applicant_status.id'), nullable=False)
    user = Column(Integer, ForeignKey('user.id'), nullable=False)

# rest of models here...

class DataAccessLayer:
    _engine: Engine = None

    @classmethod
    def db_init(cls, conn_string):
        cls._engine = create_engine(conn_string, echo=True, future=True)
        Base.metadata.create_all(bind=cls._engine)

    @classmethod 
    def session(cls) -> Session:
        return Session(cls._engine)
```

**Comment by paulespinosa** [comment]
Hi @agosmou PM has been working on statuses in a Google Doc at https://docs.google.com/document/d/1mksBNqE9hc-bAW49mdHQDImkk8f92YewtwENNpqvFHc/edit. We (devs) will need to work together with PM on defining the statuses. As of this writing, it looks like the statuses in the Google Doc are "user statuses". It's not yet clear how the relation between "User status" and "Task status" will work.

I think the Guest should not contain references to its Tasks or TaskGroups; they should be obtained via the GuestWorkflow or similar. With respect to Guest holding a reference to the GuestWorkflow, that can be done or the GuestWorkflow can be obtained via another mechanism; we won't really know which suits us until we start working with the code and gain more clarity on the domain.

`TaskStatus`,  `TaskGroupStatus`,  `TaskGroupSubProcess` are ok to hard code for starters but will need to be stored in the DB.

Try hand rolling a basic state machine to get a feel for it and gain better clarity about our needs. Thank you.

**Comment by agosmou** [comment]
cc: @tylerthome 

Thanks for this info, @paulespinosa ! Given the 'Host' items also take status, it'd be good to keep this in mind to make the code reusable for other endpoints. 

Below Iadjusted the models and made a quick run at a state machine

## Adjusted Models
```python
#models/database.py

# proposed models...

class TaskStatus(enum.Enum):
    LOCKED = "Locked"
    START = "Start"
    IN_PROGRESS = "In Progress"
    MORE_INFORMATION_NEEDED = "More Information Needed"
    COMPLETE = "Complete"


class TaskGroupStatus(enum.Enum):
    LOCKED = "Locked"
    START = "Start"
    IN_PROGRESS = "In Progress"
    MORE_INFORMATION_NEEDED = "More Information Needed"
    COMPLETE = "Complete"


class TaskGroupSubProcess(enum.Enum):
    LOCKED = "Locked"
    APPLICATION_AND_ONBOARDING_PROCESS = "Application & Onboarding Process"
    HOST_MATCHING_PROCESS = "Host Matching Process"
    MATCH_FINALIZATION_PROCESS = "Match Finalization Process"


class Guest(Base):
    __tablename__ = "guest"
    id = Column(Integer, primary_key=True, index=True)
    email = Column(String, ForeignKey("user.email"), nullable=False)
    guest_workflows = relationship(
        "GuestWorkflow", back_populates="guest", cascade="all, delete"
    )


class Task(Base):
    __tablename__ = "task"
    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=False)
    description = Column(String, nullable=False)
    task_group_id = Column(Integer, ForeignKey("task_group.id"))
    status = Column(
        Enum(TaskStatus), default=TaskStatus.LOCKED, nullable=False
    )  # from finite state machine (statechart pattern)


    # methods for state machine (statechart pattern)


```

## State Machine (State Chart Pattern)
- [Traffic Light Example](https://gist.github.com/agosmou/ed51caa68964e0354755ffbb2bdfe1bb)

pseudo-implementation of hand rolled state machine using state chart pattern

- states
- triggers
- transitions
```python
# models/state_machine.py
class TaskStateMachine:
    def __init__(self, task_id, session: Session):
        self.task = session.query(Task).get(task_id)
        self.session = session
        self.state = self.task.status

    def transition(self, trigger):
        match trigger:
            case "start":
                self.state = TaskStatus.START
            case "progress":
                self.state = TaskStatus.IN_PROGRESS
            case "need_info":
                self.state = TaskStatus.MORE_INFORMATION_NEEDED
            case "complete":
                self.state = TaskStatus.COMPLETE

        print(f"Transitioning task to {self.state}")
        self.task.status = self.state
        self.session.commit()
   
class TaskGroupStateMachine:
    def __init__(self, task_group_id, session: Session):
        self.task_group = session.query(TaskGroup).get(task_group_id)

        self.session = session
        self.state = self.task_group.status

    def transition(self, trigger):
        match trigger:
            case "start":
                self.state = TaskGroupStatus.START
            case "progress":
                self.state = TaskGroupStatus.IN_PROGRESS
            case "need_info":
                self.state = TaskGroupStatus.MORE_INFORMATION_NEEDED
            case "complete":
                self.state = TaskGroupStatus.COMPLETE

        print(f"Transitioning task group to {self.state}")
        self.task_group.status = self.state
        self.session.commit()
```

## Design
[Figma file](https://www.figma.com/file/BNWqZk8SHKbtN1nw8BB7VM/HUU-Everything-Figma-Nov-2022?type=design&node-id=1483-15519&mode=design&t=zmiPWI8j6AnuwXJW-0)

## UI Design
`~Ready for hand off - Section 2 guest dashboard`
![image](https://github.com/hackforla/HomeUniteUs/assets/73868258/6f12e775-122d-4c5b-b137-3ab6badd8237)

## Entities
`~Brainstorm - Sandbox`
![image](https://github.com/hackforla/HomeUniteUs/assets/73868258/bffbd24d-0d2b-44e3-9eba-fa42a3c9da0a)

**Comment by tylerthome** [comment]
This should be ice-boxed at least until existing MVP scope is nominally complete -- this is a requirement for compliance and traceability, but not strictly required for stakeholder demo

**Comment by tylerthome** [comment]
Moved to New Issue Approval, since this task is required for dependent specs for MVP

**Comment by lola3736** [comment]
@tylerthome lets discuss in the next Dev meeting, maybe this can be assigned to one of the team member.
